L1Trigger/VertexFinder/plugins/TPStubValueMapProducer.cc

```diff
@@ -0,0 +1,218 @@
+      stubToken_(consumes<DetSetVec>(iConfig.getParameter<edm::InputTag>("stubInputTag"))),
+      stubTruthToken_(consumes<TTStubAssMap>(iConfig.getParameter<edm::InputTag>("stubTruthInputTag"))),
+      clusterTruthToken_(consumes<TTClusterAssMap>(iConfig.getParameter<edm::InputTag>("clusterTruthInputTag"))) {
+  // Define EDM output to be written to file (if required)
+  produces<TrackingParticleCollection>();
+  produces<edm::ValueMap<l1tVertexFinder::TP>>(outputCollectionNames_[0]);
+  produces<edm::ValueMap<l1tVertexFinder::TP>>(outputCollectionNames_[1]);
+  produces<std::vector<l1tVertexFinder::TP>>(outputCollectionNames_[2]);
+}
+
+TPStubValueMapProducer::~TPStubValueMapProducer() {}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void TPStubValueMapProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+  auto vTrackingParticles = std::make_unique<TrackingParticleCollection>();
+
+  edm::Handle<TTTrackAssMap> mcTruthTTTrackHandle;
+  edm::Handle<TrackingParticleCollection> tpHandle;
+  edm::Handle<DetSetVec> ttStubHandle;
+  edm::Handle<TTStubAssMap> mcTruthTTStubHandle;
+  edm::Handle<TTClusterAssMap> mcTruthTTClusterHandle;
+  iEvent.getByToken(l1TracksMapToken_, mcTruthTTTrackHandle);
+  iEvent.getByToken(tpToken_, tpHandle);
+  iEvent.getByToken(stubToken_, ttStubHandle);
+  iEvent.getByToken(stubTruthToken_, mcTruthTTStubHandle);
+  iEvent.getByToken(clusterTruthToken_, mcTruthTTClusterHandle);
+
+  // Produce the vector of TP for the edm::Ref<TrackingParticle>->TP value map
+  unsigned int nTP = tpHandle->size();
+  auto vTPs = std::make_unique<std::vector<l1tVertexFinder::TP>>();
+  auto vTPsUse = std::make_unique<std::vector<l1tVertexFinder::TP>>();
+  vTPs->reserve(nTP);
+  vTPsUse->reserve(nTP);
+  std::set<edm::Ptr<TrackingParticle>> sTPs;
+  for (unsigned int i = 0; i < nTP; i++) {
+    TrackingParticlePtr tpPtr(tpHandle, i);
+    // Store the TrackingParticle info, using class TP to provide easy access to the most useful info.
+    l1tVertexFinder::TP tp(tpPtr, settings_);
+    // Only bother storing tp if it could be useful for tracking efficiency or fake rate measurements.
+    // Also create map relating edm::Ptr<TrackingParticle> to TP.
+    if (tp.use()) {
+      vTrackingParticles->push_back(tpHandle->at(i));
+      vTPsUse->push_back(tp);
+      sTPs.insert(tpPtr);
+    }
+    vTPs->push_back(tp);
+  }
+
+  auto vAllMatchedTPs = std::make_unique<std::vector<l1tVertexFinder::TP>>(*vTPsUse);
+  for (auto& entry : mcTruthTTTrackHandle->getTTTrackToTrackingParticleMap()) {
+    if (sTPs.count(entry.second) == 0) {
+      vAllMatchedTPs->push_back(l1tVertexFinder::TP(entry.second, settings_));
+    }
+  }
+
+  // Get the tracker geometry info needed to unpack the stub info.
+  edm::ESHandle<TrackerGeometry> trackerGeometryHandle;
+  iSetup.get<TrackerDigiGeometryRecord>().get(trackerGeometryHandle);
```

Comment:
This needs to updated following the examples here 

https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToGetDataFromES#Getting_data_from_EventSetup_wit